### DISCUSSION TITLE
build-in per-request cache

### DISCUSSION THREAD

**Discussion by lili21** [remix-run/remix#6340] [Proposals]
## Background

After introducing the Remix framework to our team and rewriting several projects using Remix, we have been happy with the end results. The developer experience has been great and we have seen significant improvements in LCP numbers, reducing them by up to 70%. However, we are constantly seeking ways to further enhance performance and have turned our attention to caching.

## HTTP Caching

We appreciate that Remix embraces web standards and have found the [Remix Run - Introduction to HTTP Caching](https://www.youtube.com/watch?v=3XkU_DXcgl0) and [CDN Caching, Static Site Generation, and Server Side Rendering](https://youtube.com/watch?v=bfLFHp7Sbkg) videos to be particularly helpful. While we were eager to try HTTP caching, the end results were not as good as we had hoped.

## Per-Page Cache Limitations

Using the headers function, implementing HTTP caching was relatively straightforward. However, we encountered issues with getting the CDN to respect the response header and cache content. Additionally, the biggest challenge we faced was that the cache was limited to the route/page level. This meant that if a page contained user-specific data, we had to move it from the server to the client.

## Client-Side Fetching

Remix provides a significant improvement to the data fetching process in terms of developer experience. However, for user-specific data, we had to fetch it on the client-side, which required us to consider loading, error handling, revalidating, and state management. and don't forget the waterfall issue.

## Proposal - Built-in Per-Request Cache
To address these limitations, we propose the implementation of a built-in per-request cache. This would allow us to cache data on a per-request basis and avoid the need to fetch user-specific data on the client-side. This would significantly reduce the complexity of our application and improve performance. By implementing this feature, we believe that Remix could provide even greater developer experience and performance improvements.

**Comment by lili21** [comment]
Another benefit is that it make the revalidate behavior more useful and performance.
For example, imagine a page/route that requires the loader to call ten different backend APIs. In addition, the page may contain various actions that update the data. In this scenario, the default revalidate behavior can be problematic as it may reload too much data, thereby reducing performance and user experience.

**Comment by addexm** [comment]
I agree that this would be a useful feature. Ideally, it would work much like Next's use of react cache. 
https://nextjs.org/docs/app/building-your-application/data-fetching/caching#react-cache

**Comment by sergiodxa** [comment]
There's a proposal to add a unified server-side cache interface https://github.com/remix-run/remix/discussions/4546, the main difference here will be that you could cache anything inside your loaders/actions, not just the loader response but parts of it, allowing for a mix on non-cached, short cached and long cached data to live together in the same response.

In my experience with Remix so far, it's better to use a server-side cache than to cache the whole response.

**Reply by addexm** [reply]
@sergiodxa A general purpose cache would definitely be useful. But I think a per-request cache presents more of a pressing need, since a general cache can be implemented easily using third-party libs.

**Reply by kiliman** [reply]
I played with a per-request cache when I first started with Remix (back before v1.0). It worked great for document requests, as we are guaranteed that even though the loaders are run in parallel, they are still part of a single request. I placed the cache object in `GetLoadContext`, and I could access it from the loaders.

However, for data requests, this is more difficult. Now the parallel loaders are *separate* requests, and there was no way to group them into the same *logical* request. I created a patch that added a header, `x-remix-request-id`, to group them together. I also included the route ids so I knew when to clean up the request cache (after all loaders were finished running).

There is still the problem if you're using serverless, as your loaders are *not* guaranteed to hit the same server.

Anyway, just some feedback on some of the issues I found.

https://github.com/kiliman/remix-playground

**Reply by addexm** [reply]
@kiliman This is super helpful. I'm going to try it out, thanks!

**Reply by kiliman** [reply]
When the RR Middleware RFC makes it in, you'll be able to set up your per-request cache there.

https://github.com/remix-run/react-router/discussions/9564

**Reply by lili21** [reply]
@kiliman Do you have any thoughts on the caching behavior of NextJS?
